File: backend/app/routes/blog.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Union
from app.models.model_user import User
from app.dao.dao_blog import BlogDAO
from app.schemas.schema_blog import BlogResponseDTO, BlogCreateDTO, BlogUpdateDTO
from app.dao.get_dao import get_blog_dao
from app.dao.get_dao import get_current_user

router = APIRouter(prefix="/blogs", tags=["blogs"])

@router.post("/", response_model=BlogResponseDTO)
async def create_blog(blog: BlogCreateDTO, current_user: User = Depends(get_current_user), dao_blog : BlogDAO = Depends(get_blog_dao)):

    try:
        new_blog = await dao_blog.create_blog(
            title=blog.title,
            content=blog.content,
            is_published = blog.is_published,
            category_id=blog.category_id,
            user_id=current_user.id
        )
        return new_blog
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail= f"Error creating blog {e}") from e

@router.get("/", response_model=List[BlogResponseDTO])
async def get_blogs(dao_blog : BlogDAO = Depends(get_blog_dao), current_user: User = Depends(get_current_user)):

    try:
        blogs = await dao_blog.get_all_blogs()
        return blogs
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error retrieving blogs") from e

@router.put("/{blog_id}", response_model=BlogResponseDTO)
async def update_blog(blog_id: int, blog: BlogUpdateDTO, current_user: User = Depends(get_current_user), dao_blog : BlogDAO = Depends(get_blog_dao)):

    try:
        existing_blog = await dao_blog.get_blogs_by_id(blog_id)
        if existing_blog is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Blog not found")
        if existing_blog.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to update this blog")

        updated_blog = await dao_blog.update_blog(
            blog_id=blog_id,
            title=blog.title,
            content=blog.content,
            is_published = blog.is_published,
            category_id=blog.category_id
        )
        return updated_blog
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail= f"Error updating blog {e}")

@router.delete("/{blog_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_blog(blog_id: int, current_user: User = Depends(get_current_user), dao_blog : BlogDAO = Depends(get_blog_dao)):

    try:
        blog = await dao_blog.get_blogs_by_id(blog_id)
        if blog is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Blog not found")
        if blog.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to delete this blog")

        await dao_blog.delete_blog(blog_id)
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error deleting blog") from e

@router.get("/{id}/", response_model = Union[BlogResponseDTO, List[BlogResponseDTO]])
async def get_blog(
    id: int, 
    get_type: str = Query(..., regex="^(USER|BLOG|CATG)$"),
    current_user: User = Depends(get_current_user),
    dao_blog : BlogDAO = Depends(get_blog_dao)
):

    try:
        if get_type == "USER":
            blogs = await dao_blog.get_blogs_by_user(id)
            if blogs is None:
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No blogs found for this user_id")
            return blogs
        
        elif get_type == "BLOG":
            blog = await dao_blog.get_blogs_by_id(id)
            if blog is None:
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No blogs found for this blog_id")
            return blog
        
        elif get_type == "CATG":
            blog = await dao_blog.get_blogs_by_category_id(id)
            if blog is None:
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No blogs found for this category_id")
            return blog

    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error retrieving blog") from e

File: frontend/src/api/blog.ts
import api from "@/services/auth";
import { Blog } from "@/interface/Blog";

// Create a new blog
export const createBlog = async (blogData: Partial<Blog>) => {
    const response = await api.post("/blogs/", blogData);
    return response.data;
  };
  
  // Get all blogs
  export const getBlogs = async () => {
    const response = await api.get("/blogs/");
    return response.data;
  };
  
  // Get a single blog by ID
  export const getBlogById = async (blogId: number) => {
    const response = await api.get(`/blogs/${blogId}/?get_type=BLOG`);
    return response.data;
  };
  
  // Update a blog
  export const updateBlog = async (blogId: number, updatedData: Partial<Blog>) => {
    const response = await api.put(`/blogs/${blogId}`, updatedData);
    return response.data;
  };
  
  // Delete a blog
  export const deleteBlog = async (blogId: number) => {
    await api.delete(`/blogs/${blogId}`);
  };
  
  // Toggle publish status of a blog
  export const togglePublishBlog = async (blogId: number) => {
    const response = await api.put(`/${blogId}/toggle-publish`);
    return response.data;
  };

  // export const getBlogbyUserId = async (id : number) =>{
  //   const response = await api.get(`/blogs/${id}`)
  //   return response.data
  // }

  export const getBlogsByUserId = async (userId?: number) => {
    if (!userId) throw new Error("User ID is not present");
    const response = await api.get(`/blogs/${userId}/?get_type=USER`);
    return response.data;
  };

  export const getBlogsByCategoryId = async (catId?: number) => {
    if (!catId) throw new Error("Category ID is not present");
    const response = await api.get(`/blogs/${catId}/?get_type=CATG`);
    return response.data;
  };

  export const searchBlogs = async (query: string) => {
    if (!query) throw new Error("Search query is required");
    const response = await api.get(`/search/blogs?q=${encodeURIComponent(query)}`);
    return response.data;
  };  